### PR TITLE
Add script to generate local javascript test coverage report

### PR DESCRIPTION
**Pull request by nospame**
## Product Description
Measuring JavaScript test coverage in HQ is a challenge for a few different reasons (described in some detail [here](https://docs.google.com/document/d/1V8n1H_3AyzTu8HnuXQekNQ_AQMU2e_h9ht1dIthoLKw/edit?usp=sharing)). This is an attempt at a script to generate a Javascript test coverage report in a limited way -- on a local dev environment, for files that are already touched by tests. The report is saved as an html page and the summary is printed to the terminal:
```
=============================== Coverage summary ===============================
Statements   : 45.26% ( 3629/8018 )
Branches     : 32.62% ( 1010/3096 )
Functions    : 35.61% ( 677/1901 )
Lines        : 45.35% ( 3598/7933 )
================================================================================

View the full coverage report at: ~/commcare-hq/coverage-js/index.html
```

## Technical Summary
[USH-3106](https://dimagi-dev.atlassian.net/browse/USH-3106)
This adds a `coverage-js` shell script that uses `git`, `nyc` (istanbul), and `grunt` commands to instrument Javascript files in place, generate coverage report under the directory `coverage-js`, and reset the file state to before Javascript was instrumented. Under the hood, `mocha-headless-chrome` is recording coverage for files touched while running tests. Because of the different pieces working together, there is a not a simple way to include coverage for files _untouched_ by tests (that is, those with 0% coverage)

Also adds a `coverage` flag to the `grunt test` task that, when included, will write coverage data to json files for each app tested (though files must be instrumented first to measure coverage, so it is only useful as part of the script).

### Safety story
The `git` commands do some lifting here to make sure local file state is saved and restored after running the report. I am new to working with shell scripts and Grunt so there may be safety concerns I have not considered.

### QA Plan
I don't think QA is applicable here.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

[USH-3106]: https://dimagi-dev.atlassian.net/browse/USH-3106?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by nospame** [comment]
Marking this as ready for review, but flagging I'm fully open to changes to structure or approach based on reviewers' thoughts.

File: Gruntfile.js [hunk 1]
```diff
@@ -85,6 +86,12 @@ module.exports = function(grunt) {
             if (data.result.failures.length) {
                 failures[currentApp] = data.result.failures;
             }
+            if (grunt.option('coverage')) {
+                var coverageDir = './coverage-js/',
+                    filePath = coverageDir + currentApp.replace(/\//g, '-') + '.json';
+                if (!fs.existsSync(coverageDir)) { fs.mkdir(coverageDir, err => console.log(err)); }
```
Comment (Robert-Costello) [inline Gruntfile.js]: Is there cleanup of `coverageDir` at some point or is the idea to have this available after the cover report script is run?
Is `coverageDir` overwritten each time you run the script?

Comment (nospame) [reply Gruntfile.js]: The HTML version of the report ends up that directory and will want to stick around for sure. The json coverage files that are generated here could be useful for other tools but aren't strictly necessary to keep around. They'll get overwritten each time it runs, unless for some reason an app gets removed from the list of test apps (in which case that json file would just stay indefinitely). 

I could see a pre-run cleanup to make sure everything that makes it into the report is current -- something really simple like delete the directory if it exists, then run the script. Would that make sense?

Comment (Robert-Costello) [reply Gruntfile.js]: Yeah, I think some pre-cleanup would be good for keeping the directory concise and all of the files contained in it relevant to the current state of the js code.

**Comment by nospame** [comment]
> I took a shot at running the script locally and I get `ENOENT: no such file or directory, stat './coverage-js'` Creating `coverage-js` manually fixes this. (Maybe some order of operations issue that isn't obvious?)
> 
> When I run it again I get this:
> 
> Statements : Unknown% ( 0/0 ) Branches : Unknown% ( 0/0 ) Functions : Unknown% ( 0/0 ) Lines : Unknown% ( 0/0 )
> 
> and `codecoverage.json` = {}
> 
> looking at the code it's not clear to me why, any thougths?

@Robert-Costello 
The cause of that `ENOENT` isn't immediately clear to me -- removing the directory and running the script works fine in my environment, but I imagine it could be either a permissions issue or a timing issue between `existsSync` and `mkdir`. 

As for the blank coverage report, that's an odd one. It seems like `nyc` is running just fine, but maybe the instrumenting step is failing if you're getting blank json files (which would indicate no coverage recorded). Or possibly the coverage data isn't coming out of Grunt/mocha-headless-chrome correctly. Is it generating coverage files at all (`app_manager.json`, `case_importer.json`, etc)?

In running this again I did come across another issue where the git history can end up in a weird state, since it's currently assuming there _are_ changes to stash, but if there are none it ultimately resets 1 commit too far back, so will look at a fix there.

**Comment by Robert-Costello** [comment]
>but if there are none it ultimately resets 1 commit too far back, so will look at a fix there.

Good catch. 

> Is it generating coverage files at all (app_manager.json, case_importer.json, etc)?

I'm not getting `app_manager.json` or `case_importer.json`. The only json generated is the empty `case_importer.json`

**Comment by Robert-Costello** [comment]
looks great and very user friendly!

**Review by Robert-Costello — APPROVED** [review]
One additional thing that would be helpful is some documentation similar to what's in the description of this PR. Maybe in DEV_SETUP.md

**Review by Robert-Costello — APPROVED** [review]

**Pull request merged**